### ISSUE TITLE
Cases for verification for cosign 2.0

### ISSUE THREAD

**Issue by asraa** [sigstore/cosign#2466]
**Description**

@znewman01 @haydentherapper these are coming up in my refactor, and I want to enumerate them and we can close this out when we know we have testing coverage for these cases.

@priyawadhwa @hectorj2f 

We currently have the following options available to users:
* Offline
* TSA bundle
* (Rekor) bundle
* Key
* Cert
* RekorClient
* TSAClient
* SkipTlogVerify

I was hoping it'd be clear what the behavior was in the document, but let's just spec it out here instead.

Proposal:
* Rename SkipTlogVerify to RequireTlogVerify (default on)
* Remove the CheckOpt for TSAClient. It's not used for verification, and only used as a toggle for whether we should verify against the TSA.

This will make the flow as such:

1. If a certificate is provided, verify it's chain up to the trust roots. Verify other properties designed from CheckOpts (issuer etc)
2. Verify the cryptographic signature.
3. Check any claims because now we can decode the payload.
4. Rekor. 
  a.. [offline] If a (Rekor) bundle is provided, verify it. If fails, exit. If succeeds, extract tlogVerificationTime.
  b. If we requireTlogVerification and a bundle wasn't present:
     i. If we are offline, fail because we didn't have a bundle.
    ii. We are now definitely online (not offline) and a bundle wasn't provided and we required tlog verification, so perform online Rekor verification. Extract tlogVerificationTime.
5. TSA:
  a. [offline] If a TSA bundle is present, verify it. If fails, exit. If succeeds, extract tsaVerificationTime. (Not that I did NOT say `if TSAClient was set @hectorj2f -- the TSAClient is NOT used in `VerifyTSABundle` anyway)
6. Check the certificate verification time against the tlogVerificationTime (if present) and the tsaVerificationTime (if present). Ensure that at least one was checked to ensure signature validity (in case we didn't have a TSA bundle, didnt have a rekor bundle, and didnt require rekor validation).

Thoughts? We don't get to force online rekor proofs if a rekor bundle is present, but such is life.

**Comment by haydentherapper** [comment]
Edit: I guess if you have to pass in the cert chain for which TSA you trust, this is sufficient for saying which TSAs you trust (same with if the tsa chain comes from TUF).

**Comment by asraa** [comment]
> To confirm, offline flag is set to true by default?

I would actually set it to false... and people opt in to offline.

> What do you meant if "did not pass"? As in, there's no bundle present?

Good question. Yes: as in, there's no bundle. I'll word that better in the original issue. If it was present and it fails verification we exit.

> Should you check this before a and b so you don't actually do lookups if it's not required?

I think so, updated.

**Comment by haydentherapper** [comment]
> I would actually set it to false... and people opt in to offline.

That makes sense. So if the bundle is present and passes, we don't check online still (unless we add a flag to force online?)

> I think so, updated. EDIT: update: I don't think so. If you had the bundle, just verify it.

I think we need to be clear on the behavior of the RequireTlogVerify flag then. So if you have a bundle, and it fails, then you'll exit, which is reasonable. What about if you a) have no bundle, b) don't set offline, c) set RequireTlogVerify=false? Per the proposed logic, this would still make a network call

**Comment by asraa** [comment]
> What about if you a) have no bundle, b) don't set offline, c) set RequireTlogVerify=false? Per the proposed logic, this would still make a network call

Check the updated logic! In the new flow it would not.

**Comment by haydentherapper** [comment]
Updated flow LGTM! Couple more questions:

> TSA:
a. [offline] If a TSA bundle is present, verify it. If fails, exit. If succeeds, extract tsaVerificationTime. (Not that I did NOT say if TSAClient was set @hectorj2f -- the TSAClient is NOT used in VerifyTSABundle` anyway)

What if someone removes the timestamp from the OCI annotation? For Rekor, it would fall back to online. There's no fallback here (i guess the fallback is using the rekor timestamp, so maybe that's fine?) 

> Check the certificate verification time against the tlogVerificationTime (if present) and the tsaVerificationTime. Ensure that at least one was checked to ensure signature validity.

What do we do if one fails? Print a warning? I think we should require that both succeed, and if there's pushback, add policy flags to dictate which you trust (also solvable via delegations in the future)

**Comment by znewman01** [comment]
I think these are all fine as incremental improvements. If I'm allowed to dream big:

--------------------------------------------------------------------------------

My overall philosophy here is to *have flags to specify really clearly* what's going on, and then *set sensible defaults* so most people don't have to think about them. I think "boolean" is the wrong type for many of these flags, which can interact in complicated ways.

So here's a proposal:

- Flags
  - What Sigstore infra to use?
    - Where do the keys come from?:
      - `--tuf-root=URL` (default: Sigstore public good)
      - `--fulcio-root` (replaces `--certificate-chain`; default: comes from TUF root)
      - `--rekor-root` (default: comes from TUF root)
      - `--ct-log-root` (default: comes from TUF root)
    - `--rekor-url=URL`, `--fulcio-url=URL`, `--tsa-url=URL` (defaults: URLs provided from the TUF root)
  - How should Cosign actually verify?
    - `--rekor-verification={offline,fallback,online,insecure}` (default `fallback`)
    - `--sct-verification={offline,fallback,online,insecure}` (default `fallback`; replaces `--enforce-sct` flag)
    - `--tsa-verification={offline,fallback,online}` (default `fallback`)
  - Verification material:
    - `DATA` (positional): file name for `verify-blob`, `gcr.io/foo:latest` for `verify`
    - `--bundle` (the new bundle format)
       - Can replace all of these:
         - `--rekor-bundle=FILE`
         - `--certificate=FILE`
         - `--signature=FILE`
      - For *OCI only*, default is fallback to fetching from OCI.
    - `--offline` flag doesn't affect fetching verification material from OCI registry!
  - Identity flags: 
    - Require one of:
      - `--certificate-issuer` AND `--certificate-identity` (optionally: also `--certificate-github-*`)
      - `--key`
      - `--sk` (and optional `--slot`, which should be renamed to `--sk-slot`)
  - `--offline`: force all of the flags above to `offline` (except probably fetching the image from the OCI registry)
- If anything is *explicitly* provided (e.g., via a flag) and doesn't validate, fail.
- If anything is *implicitly* provided (e.g., via a "fallback", or via the bundle) and doesn't validate, warn ("found TSA in the Cosign annotation but it fails").

I believe this accomplishes the following goals:

- Doesn't exhibit "adding more information makes verification fail" behavior for implicit 
- Only makes necessary network calls

-----------------------------------------------------------------------------------------

One point I'd like to respond to:

> > > Check the certificate verification time against the tlogVerificationTime (if present) and the tsaVerificationTime. Ensure that at least one was checked to ensure signature validity.
> > 
> > What do we do if one fails? Print a warning? I think we should require that both succeed, and if there's pushback, add policy flags to dictate which you trust (also solvable via delegations in the future)
>
> Agree, this checks both, and requires at least one to have been checked (just in case no TSA is provided, and no bundle was provided)

(Related also to "Remove the CheckOpt for TSAClient. It's not used for verification, and only used as a toggle for whether we should verify against the TSA.")

I'd like folks to be more explicit about what they're expecting here (with a reasonable default). This strikes me as a potential source of confusion.

**Comment by asraa** [comment]
AMAZING. 

I like the idea of specifying with an enum over a boolean for the fallbacks/etc

As an incremental improvement, maybe we can at least adopt that for the rekor verification side for 2.0?

Likewise for the identity flags, this is something we can do with our existing logic already.

**Comment by vaikas** [comment]
Regarding:
```
--rekor-verification={offline,fallback,online,insecure}
```

And a couple of other places that talk about fallback, would be good to explicitly specify what the behaviour there is. I poked around a little bit and seems like it may be (just double-checking if it's right):
try offline, if there's no bundle then go check online. Is that right?

I think I also saw somewhere that we may want something that verifies both: offline, and online? If that's the case then it seems like we may need more field values to specify that.

I think that's (fallback) what @asraa is describing here:
```
Rekor.
a.. [offline] If a (Rekor) bundle is provided, verify it. If fails, exit. If succeeds, extract tlogVerificationTime.
b. If we requireTlogVerification and a bundle wasn't present:
i. If we are offline, fail because we didn't have a bundle.
ii. We are now definitely online (not offline) and a bundle wasn't provided and we required tlog verification, so perform online Rekor verification. Extract tlogVerificationTime.
```

Also `insecure` seems interesting. Does it mean it's doing some verification but in an insecure manner (http vs. https for example), or that there's no verification done (in which case it should imho be called `none`).

**Comment by znewman01** [comment]
> Regarding:
> 
> ```
> --rekor-verification={offline,fallback,online,insecure}
> ```
> 
> And a couple of other places that talk about fallback, would be good to explicitly specify what the behaviour there is. I poked around a little bit and seems like it may be (just double-checking if it's right): try offline, if there's no bundle then go check online. Is that right?

Yes, that's what I had in mind.
> 
> I think I also saw somewhere that we may want something that verifies both: offline, and online? If that's the case then it seems like we may need more field values to specify that.

`both`? Not sure under what circumstances we'd need that, though. IMO `online` is a strictly better check than `offline`. (I could be totally wrong here) 

> 
> I think that's (fallback) what @asraa is describing here:
> 
> ```
> Rekor.
> a.. [offline] If a (Rekor) bundle is provided, verify it. If fails, exit. If succeeds, extract tlogVerificationTime.
> b. If we requireTlogVerification and a bundle wasn't present:
> i. If we are offline, fail because we didn't have a bundle.
> ii. We are now definitely online (not offline) and a bundle wasn't provided and we required tlog verification, so perform online Rekor verification. Extract tlogVerificationTime.
> ```

IIUC that is an attempt at trying to turn a boolean `requireTlogVerification` into  pseudocode, and it's over-complicated because it's not terribly clear what the users want in that case. With the enum cases it's more like:

- `insecure-none`: `return true`
- `offline`: `return bundle != nil && verifyBundle(bundle)
- `online`: `return verifyOnline()`
- `fallback`: `if bundle != nil { return verifyBundle(bundle) } else { return verifyOnline() }`

> Also `insecure` seems interesting. Does it mean it's doing some verification but in an insecure manner (http vs. https for example), or that there's no verification done (in which case it should imho be called `none`).

The reason we wanted to write `insecure` is to make users acknowledge that it's insecure. Maybe `insecure-none`?

**Comment by vaikas** [comment]
re: `both` I could've sworn that I saw something that @haydentherapper said somewhere that there would be some cases where one would want to do the offline as well as online. If I'm remembering wrong, let's drop it :)

re: `insecure` I guess nothing says like insecure like 'not verifying anything' :)

**Comment by haydentherapper** [comment]
There's the case where I have a bundle, but I want to do an online lookup (because online inclusion proofs will always give you stronger guarantees than offline promises). I guess the question is, should we just ignore any bundle presented if you want to do an online lookup, or always verify what's given in addition to an online lookup?

**Comment by znewman01** [comment]
> There's the case where I have a bundle, but I want to do an online lookup (because online inclusion proofs will always give you stronger guarantees than offline promises). I guess the question is, should we just ignore any bundle presented if you want to do an online lookup, or always verify what's given in addition to an online lookup?

IMO if you want online, you just want online even if you have a bundle on hand because online >> offline.

But I won't die on this hill—`both` is very clear and easy to implement.

**Comment by asraa** [comment]
Should we close this? I think we have the testcases in, at least in blobs.

**Comment by hectorj2f** [comment]
I realized I totally missed this issue :/. I also like the idea of specifying with an enum over a boolean for the fallbacks 🌟 .

I have couple of comments. The TSAClient (or what I understand today as TSACertificates) are currently used for TSA verification. I am not sure if you referred to some old legacy code we initially added where we had the tsa client in the CheckOpts.

```
tsaverification.VerifyOpts{
			TSACertificate: co.TSACertificate,
			Intermediates:  co.TSAIntermediateCertificates,
			Roots:          co.TSARootCertificates,
		})
```

Also, I recently renamed `InsecureSkipTlogVerify` to `InsecureIgnoreTlog`. I hope we are okay with this new name instead of the `RequireTlogVerify`.
